### PR TITLE
nhost: init at 1.5.2

### LINKED ISSUE #245798: Package request: nhost-cli
**Project description**

nhost is an open source firebase alternative (postgres graphql hasura authentication and storage)

**Metadata**

* homepage URL: https://nhost.io/
* source URL: https://github.com/nhost/cli
* license: mit
* platforms: mac, linux, windows wsl,

### PR DESCRIPTION
**Pull request by andr-ec**
## Description of changes
Adds the Nhost CLI tool and closes https://github.com/NixOS/nixpkgs/issues/245798

this package requires `  virtualisation.docker.enable = true;` anything I should do with that? 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/web/nhost/default.nix
Comment (cafkafk) [inline pkgs/development/web/nhost/default.nix:4]:
```suggestion
{ 
  buildGoModule
, fetchFromGitHub
, lib
}:
```

This makes diffs more readable.

File: pkgs/development/web/nhost/default.nix
Comment (cafkafk) [inline pkgs/development/web/nhost/default.nix:27]: You need to add yourself to the [maintainers file.](https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix#L1)

File: pkgs/development/web/nhost/default.nix
Comment (cafkafk) [inline pkgs/development/web/nhost/default.nix:24]: This should not end with a period.

See: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes

**Review by cafkafk — CHANGES_REQUESTED** [review]
This looks okay, just needs some small changes :+1: 

Of course, also needs to fix issues ofborg mentions.

File: pkgs/development/web/nhost/default.nix
Comment (cafkafk) [reply pkgs/development/web/nhost/default.nix:27]: Btw this might be what's causing “This PR does not cleanly list package outputs after merging.”

File: pkgs/development/web/nhost/default.nix
Comment (andr-ec) [reply pkgs/development/web/nhost/default.nix]: done, I just copied another example

File: pkgs/development/web/nhost/default.nix
Comment (andr-ec) [reply pkgs/development/web/nhost/default.nix:28]: done, thanks!

File: pkgs/development/web/nhost/default.nix
Comment (andr-ec) [reply pkgs/development/web/nhost/default.nix]: done

**Comment by cafkafk** [comment]
As ofborg says:

```
trace: maintainer andreasfelix (line 974) should be listed after andir, not after andr-ec (line 968)
error: assertion '(errors == 0)' failed
```
There also seems to be an issue with a newline.

Ideally, I think it would be best to squash all these commits into a single commit `nhost: init at 1.5.2`, but then if that builds this should be ready!

**Comment by andr-ec** [comment]
Looks like it all built, thanks for the help!

**Comment by cafkafk** [comment]
> Looks like it all built, thanks for the help!

nw!

Now, I think if you want a better chance of being merged, you'll have to squash all these commits into a single commit titled `nhost: init at 1.5.2`. 

You'll probably be using `git rebase -i <commit before your first>`, you can read more about it [here](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History).

File: pkgs/development/web/nhost/default.nix
Comment (drupol) [inline pkgs/development/web/nhost/default.nix:15]: Prefer `hash = ...` to `sha256 = ...` Both currently work, but the former will make it easier if the algorithm ever needs to change.

Since there could be different hashing algorithms in use, this make sense.

File: pkgs/development/web/nhost/default.nix
Comment (drupol) [inline pkgs/development/web/nhost/default.nix:18]: use `vendorHash`

File: pkgs/development/web/nhost/default.nix
Comment (drupol) [inline pkgs/development/web/nhost/default.nix:24]: Maybe you should add a `meta.mainProgram` attribute.

File: pkgs/top-level/all-packages.nix
Comment (drupol) [inline pkgs/top-level/all-packages.nix:10821]: Please try to add this at the right place. The letter `h` comes after the letter `e`.

File: pkgs/development/web/nhost/default.nix
Comment (drupol) [inline pkgs/development/web/nhost/default.nix:2]:
```suggestion
{ buildGoModule
```

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your contribution.

I've left some comments, left me know if this is clear enough.

Remember that this PR should only contains 2 commits in the end with specific commit messages:

- 1 commit for adding you as maintainer: `maintainers: add andr-ec`
- 1 commit for the rest: `nhost: init at 1.5.2`

File: pkgs/development/web/nhost/default.nix
Comment (drupol) [reply pkgs/development/web/nhost/default.nix]: @cafkafk Do you use a tool to format the file? When using `nixpkgs-fmt`, I have the original PR formatting and this is the one that I usually recommend. It would be nice if reviewers could be on the same wavelength on this or else contributors might go crazy.

Comment (cafkafk) [reply pkgs/development/web/nhost/default.nix]: > It would be nice if reviewers could be on the same wavelength on this or else contributors might go crazy.

Agreed!

The style I'm using is based on what's been recommended to me on reviews I've gotten, I tried using nixpkgs-fmt, but reviewers nearly always had problems with it. 

I'm only recommending this because it's what my experience suggests is idiomatic, and what I would expect other reviewers would mention if I didn't. I may be wrong however!

It is a problem that there isn't an unified style guide for nixpkgs, but what I'm arguing for aligns more with one popular style guide from e.g. [alejandra](https://github.com/kamadorueda/alejandra) which quite nicely argues for this.

Allow me to copy-paste it here:

# from alejandra style guide

✅ Good:

```nix
{
  mkDerivation,
  lib,
  fetchurl,
  ...
} @ attrs:
  stdenv.mkDerivation # ...
```

- Adding an argument produces a minimal diff
  (including the first and last elements):

  ```patch
    mkDerivation,
    lib,
    fetchurl,
  + google-chrome-stable,
  ```

- Removing an argument produces a minimal diff
  (including the first and last elements):

  ```patch
    mkDerivation,
  - lib,
    fetchurl,
  ```

- The comma at the end is consistent with _Let-In_, and _Map_,
  where the separator goes after the element
  instead of at the beginning.

❌ Bad:



```nix
{ lib
, mkDerivation
, fetchurl
, ...
} @ attrs:
stdenv.mkDerivation # ...
```

- Removing the first element
  produces a diff in two elements:

  ```diff
  - { lib
  - , mkDerivation
  + { mkDerivation
    , fetchurl
    , ...
    } @ attrs:
    stdenv.mkDerivation # ...
  ```

- Documenting the first argument creates an inconsistency
  between the way arguments start:

  ```nix
  {
    # Lorem Ipsum
    lib
  , mkDerivation
  , fetchurl
  , ...
  } @ attrs:
  stdenv.mkDerivation # ...
  ```

- This is not consistent with _Let-In_, and _Map_,
  where the separator goes after the element
  instead of at the beginning.
- It ruins "folding by indentation" modes
  on Vim, Neovim, VSCode, and other major code editors,
  because the data-structure has the same indentation
  as the opening bracket.

❌ Bad:



```nix
{ mkDerivation, lib, fetchurl, ... }@attrs: stdenv.mkDerivation # ...
```

- One-liners are unreadable.

❌ Bad:



```nix
{ mkDerivation, lib, fetchurl, extra-cmake-modules, kdoctools, wrapGAppsHook
, karchive, kconfig, kcrash, kguiaddons, kinit, kparts, kwind, ... }@attrs:
stdenv.mkDerivation # ...
```

- It's hard to tell this destructuring has an ellipsis (`...`) at a first glance,
  because it's mixed with the other arguments.
- Moving elements becomes harder
  than a simple whole-line movement.
  (Moving a whole line is normally a keyboard-shortcut
  or command in major code editors).
- Excessively compact:
  adding, removing, or editing an argument
  produces a diff in more than one argument.
- `}@attrs` is not intuitive
  with the rules of written english,
  where you add whitespace
  after the end of the previous phrase
  (`phrase. Other phrase`).

Comment (cafkafk) [reply pkgs/development/web/nhost/default.nix]: Of course either way isn't blocking to me, it's more an improvement suggestion, whatever way @andr-ec goes with that pleases you is fine with me!

Comment (drupol) [reply pkgs/development/web/nhost/default.nix]: Impressive. Maybe I should use your syntax. I'll think about it. Thanks for the detailed explanation !!!

Comment (AndersonTorres) [reply pkgs/development/web/nhost/default.nix]: > Removing the first element produces a diff in two elements:

`lib` and `stdenv` were always the first elements. Then, at least in Nixpkgs, this particular case never happens.

And I believe this is too ugly and injustifiable for the input attrset.

**Pull request closed without merging**
